state_machine/drivers/pycubedmini/lib/opt3001.py

```diff
@@ -0,0 +1,88 @@
+"""
+CircuitPython driver for the OPT3001 ambient light sensor
+M. Holliday
+
+     GND  3V3  SDA  SCL
+ADDR 0x44 0x45 0x46 0x47
+
+from opt3001 import OPT3001
+s = OPT3001(cubesat.i2c1, 0x45)
+"""
+
+import time
+from micropython import const
+from adafruit_bus_device.i2c_device import I2CDevice
+from adafruit_register.i2c_bits import RWBits
+from adafruit_register.i2c_bit import ROBit, RWBit
+
+RESULT = const(0x00)
+CONFIGURATION = const(0x01)
+
+class OPT3001:
+    """OPT3001 Sun Sensor Driver"""
+
+    # RWBits(num_bits, register_address, lowest_bit, register_width=1, lsb_first=True)
+
+    # See Table 10 in datasheet for Configuration Register settings
+    mode         = RWBits(2, CONFIGURATION, 9, register_width=2, lsb_first=False)
+    range_number = RWBits(4, CONFIGURATION, 12, register_width=2, lsb_first=False)
+    conv_800ms   = RWBit(CONFIGURATION, 11, register_width=2, lsb_first=False)
+    rdy   = ROBit(CONFIGURATION, 7, register_width=2, lsb_first=False)
+
+    def __init__(self, i2c_bus, address=0x44):
+        """Initialize and Configure the Light Sensor Driver"""
+        self.i2c_device = I2CDevice(i2c_bus, address)
+        # Initialize a fixed buffer to read and write from
+        self.buf = bytearray(2)
+
+        self.read_u16(0x7F)  # DEVICE_ID
+        if b'0\x01' not in self.buf:
+            print(f'device id: {self.buf}')
+            raise Exception("ERROR: Could not read correct device ID from bus provided")
+
+        self.read_u16(0x7E)  # MANUFACTURER_ID
+        if b'TI' not in self.buf:
+            print(f'manufacturer id: {self.buf}')
+            raise Exception("ERROR: Could not read correct manufacturer ID from bus provided")
+
+        # CONFIGURE
+        self.range_number = 0b1100  # auto ranging
+        self.conv_800ms   = True   # 800ms conversion time
+        self.mode = 0b11           # continuous conversions
+
+    def read_u16(self, addr):
+        self.buf[0] = addr
+        with self.i2c_device as i2c:
+            i2c.write_then_readinto(self.buf, self.buf, out_end=1, in_start=0)
+
+    def lux(self):
+        """LUX value of sun-sensor. Could have delay up to chosen conversion time (800 or 100ms)"""
+        # wait until the lux measurement is ready
+        _t = time.monotonic() + 1.1
+        while time.monotonic() < _t:
```

Comment:
If this could be up to 800ms this loop could introduce some serious delay. I think this might need to be async? @thetazero any input here?

Reply:
Even 100ms is unacceptable for this to block. 

Are there other options for the data returned by the sun sensor that are faster?

Reply:
This should definitely be async

Reply:
There are no faster options, according to the spec sheet the fastest setting is on average 100ms and can be anywhere in between 90 ms and 110 ms.

However, I'm not sure making this function async would do anything to help us as it returns a value we need. For example, In the case of the sun_vector property we would need to wait for the lux property to return a value in order to return our result, so we will still experience the same delay, despite lux being async.